move/squid/sources/squid/deepbook_v3.move

```diff
@@ -0,0 +1,215 @@
+    public fun estimate<B, Q>(self: &mut SwapInfo, pool: &Pool<B, Q>, clock: &Clock) {
+        let data = self.get_data_swapping();
+        if (data.length() == 0) return;
+        let swap_data = peel_swap_data(data);
+
+        assert!(swap_data.swap_type == SWAP_TYPE, EWrongSwapType);
+        assert!(swap_data.pool_id == object::id_address(pool), EWrongPool);
+        assert!(
+            &swap_data.base_type == &type_name::get<B>().into_string(),
+            EWrongCoinType,
+        );
+        assert!(
+            &swap_data.quote_type == &type_name::get<Q>().into_string(),
+            EWrongCoinType,
+        );
+
+        if (swap_data.has_base) {
+            let (base_out, quote_out, _deep_required) = pool.get_quote_quantity_out(
+                self.coin_bag().get_estimate<B>(),
+                clock
+            );
+            if (swap_data.min_output > quote_out) {
+                self.skip_swap();
+                return
+            };
+            if (!swap_data.should_sweep) self.coin_bag().store_estimate<B>(base_out);
+            self.coin_bag().store_estimate<Q>(quote_out);
+        } else {
+            let (base_out, quote_out, _deep_required) = pool.get_base_quantity_out(
+                self.coin_bag().get_estimate<Q>(),
+                clock
+            );
+            if (swap_data.min_output > base_out) {
+                self.skip_swap();
+                return
+            };
+            if (!swap_data.should_sweep) self.coin_bag().store_estimate<Q>(quote_out);
+            self.coin_bag().store_estimate<B>(base_out);
+        }
+    }
+
+    /// Perform a swap. First, check how much DEEP is required to perform the swap.
+    /// Then, get that amount of DEEP from squid. Use it to swap the entire amount
+    /// of the base or quote currency. Store any remaining DEEP back in squid.
+    public fun swap<B, Q>(
+        self: &mut SwapInfo,
+        pool: &mut Pool<B, Q>,
+        squid: &mut Squid,
+        clock: &Clock,
+        ctx: &mut TxContext,
+    ) {
+        let data = self.get_data_swapping();
+        if (data.length() == 0) return;
+        let swap_data = peel_swap_data(data);
+
+        assert!(swap_data.swap_type == SWAP_TYPE, EWrongSwapType);
+        assert!(swap_data.pool_id == object::id_address(pool), EWrongPool);
+        assert!(
+            &swap_data.base_type == &type_name::get<B>().into_string(),
+            EWrongCoinType,
+        );
+        assert!(
+            &swap_data.quote_type == &type_name::get<Q>().into_string(),
+            EWrongCoinType,
+        );
+
+        if (swap_data.has_base) {
+            let (_, _, deep_required) = pool.get_quote_quantity_out(
+                self.coin_bag().get_estimate<B>(),
+                clock
+            );
+            let deep_in = squid.coin_bag().get_exact_balance<DEEP>(deep_required).destroy_some().into_coin(ctx);
+            let base_in = self.coin_bag().get_balance<B>().destroy_some().into_coin(ctx);
```

Comment:
Similarly, should we take out 10bps worth of base_in to compensate for deep_in? Would we transfer that to squid?

Reply:
That is what we discussed (Fig said maybe take a custom amount but I am now realising that anyone could craft custom transactions to use Squid's DEEP without giving them anything, so the amount held should be constant.